File: src/services/liveService.js
/**
 * Live / near-real-time situational data service (frontend).
 *
 * Polls the FastAPI backend `/api/live/*` endpoints. Because the Vercel deployment is
 * a pure static build (no serverless backend), the service transparently falls back to
 * a built-in DemoProvider when the backend is unreachable, and ALWAYS labels the data
 * truthfully (LIVE / DEMO / STALE) so the UI never overclaims.
 */
import { WAYANAD_DATA } from '../data/wayanadData.js';

const API_BASE = (import.meta.env.VITE_API_BASE || '').replace(/\/$/, '');

// ── Configuration (adjustable via Vite env) ────────────────────────────────
const CONFIG = {
  refreshIntervalSeconds: Number(import.meta.env.VITE_LIVE_REFRESH_SECONDS || 60),
  staleAfterMinutes: Number(import.meta.env.VITE_LIVE_STALE_MINUTES || 15),
  backendTimeoutMs: 8000,
};

function isoNow() {
  return new Date().toISOString();
}

function fmtClock(iso) {
  if (!iso) return '—';
  const d = new Date(iso);
  if (isNaN(d.getTime())) return '—';
  return d.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit', second: '2-digit' });
}

/**
 * Client-side demo provider — mirrors the backend DemoProvider so the live map is
 * fully demonstrable without a backend. Clearly labelled SIMULATED.
 */
class DemoProvider {
  source = 'DEMO';

  getWeather() {
    const minute = new Date().getMinutes();
    const rain = Math.round((84 + (minute % 7) * 1.5) * 10) / 10;
    return {
      source: 'DEMO',
      timestamp: isoNow(),
      status: 'DEMO',
      data_age_minutes: 0,
      region: 'Wayanad, Kerala',
      station_id: 'DEMO-MEPPADI-AWS',
      rainfall_24h_mm: rain,
      rainfall_now_mm: Math.round((6 + (minute % 5) * 1.2) * 10) / 10,
      temperature_c: Math.round((23.5 + (minute % 4) * 0.3) * 10) / 10,
      humidity_pct: 88 + (minute % 6),
      wind_kph: Math.round((9 + (minute % 5)) * 10) / 10,
      forecast: 'Moderate to heavy rainfall likely over next 24 hours.',
    };
  }

  getAlerts(weather) {
    const rain = weather && weather.rainfall_24h_mm;
    const alerts = [];
    let severity = 'LOW';
    if (rain != null && rain >= 140) {
      severity = 'HIGH';
      alerts.push({
        id: 'imd-heavy-rain',
        type: 'HEAVY RAIN WARNING',
        severity: 'HIGH',
        area: 'Wayanad',
        issued: weather.timestamp,
        message: '24h rainfall crossing 140mm threshold — high-risk slope zones.',
        source: 'DEMO',
      });
    } else if (rain != null && rain >= 90) {
      severity = 'MODERATE';
      alerts.push({
        id: 'imd-rain-advisory',
        type: 'RAIN ADVISORY',
        severity: 'MODERATE',
        area: 'Wayanad',
        issued: weather.timestamp,
        message: 'Elevated rainfall observed — monitor landslide-prone settlements.',
        source: 'DEMO',
      });
    }
    return { alerts, count: alerts.length, severity, status: 'DEMO', source: 'DEMO', timestamp: isoNow() };
  }

  getRiskMap(weather) {
    const rain = weather && weather.rainfall_24h_mm;
    let rainFactor = 0;
    if (rain != null) {
      if (rain >= 140) rainFactor = 0.35;
      else if (rain >= 90) rainFactor = 0.2;
      else if (rain >= 50) rainFactor = 0.1;
    }
    const parseRain = (s) => {
      const m = /([\d.]+)/.exec(s.rainfall24h || '');
      return m ? parseFloat(m[1]) : 55;
    };
    const features = WAYANAD_DATA.settlements.map((s) => {
      const rainfallRisk = Math.min(100, parseRain(s) + rainFactor * 100);
      const delta = Math.round(rainFactor * 12); // modest live uplift for demo
      const score = Math.min(100, s.riskScore * 10 + delta);
      const level = score >= 80 ? 'CRITICAL' : score >= 60 ? 'HIGH' : 'MODERATE';
      return {
        type: 'Feature',
        properties: {
          id: s.id,
          name: s.name,
          risk_score: Math.round(score),
          risk_level: level,
          households: s.displacedFamilies,
          priority_households: s.demographics.elderlyAndDisabled + s.demographics.childrenUnder10,
          previous_risk_score: Math.round(s.riskScore * 10),
          previous_risk_level: s.riskLevel,
          rainfall_24h_mm: rain,
          trend: score > s.riskScore * 10 + 1 ? 'increasing' : 'stable',
        },
        geometry: { type: 'Point', coordinates: [s.coordinates[1], s.coordinates[0]] },
      };
    });
    return {
      updated_at: isoNow(),
      source: 'DEMO',
      status: 'DEMO',
      data_age_minutes: 0,
      rainfall_24h_mm: rain,
      rain_impact_factor: rainFactor,
      type: 'FeatureCollection',
      features,
    };
  }
}

class LiveService {
  constructor() {
    this.demo = new DemoProvider();
    this.timer = null;
    this.running = false;
    this.lastGoodWeather = null;
    this.listeners = new Set();
    this.data = this._initial();
  }

  _initial() {
    return {
      mode: 'DEMO',
      status: 'DEMO',
      weather: null,
      alerts: { alerts: [], count: 0, severity: 'LOW' },
      riskMap: null,
      statusSnapshot: null,
      lastUpdate: null,
      nextUpdate: null,
      stale: false,
      backendReachable: false,
      error: null,
    };
  }

  subscribe(fn) {
    this.listeners.add(fn);
    return () => this.listeners.delete(fn);
  }

  _emit() {
    this.listeners.forEach((fn) => fn(this.data));
  }

  async _fetchJson(path) {
    const url = `${API_BASE}${path}`;
    const controller = new AbortController();
    const to = setTimeout(() => controller.abort(), CONFIG.backendTimeoutMs);
    try {
      const res = await fetch(url, { signal: controller.signal });
      if (!res.ok) throw new Error(`HTTP ${res.status}`);
      return await res.json();
    } finally {
      clearTimeout(to);
    }
  }

  async _refresh() {
    const now = Date.now();
    let weather = null;
    let riskMap = null;
    let alerts = null;
    let statusSnapshot = null;
    let backendReachable = false;

    try {
      const results = await Promise.allSettled([
        this._fetchJson('/api/live/weather'),
        this._fetchJson('/api/live/alerts'),
        this._fetchJson('/api/live/risk-map'),
        this._fetchJson('/api/live/status'),
      ]);
      if (results[0].status === 'fulfilled' && results[0].value) {
        weather = results[0].value;
        this.lastGoodWeather = weather;
        backendReachable = true;
        alerts = results[1].status === 'fulfilled' ? results[1].value : null;
        riskMap = results[2].status === 'fulfilled' ? results[2].value : null;
        statusSnapshot = results[3].status === 'fulfilled' ? results[3].value : null;
      } else {
        throw new Error('Backend unreachable');
      }
    } catch (e) {
      backendReachable = false;
      // Fallback: use last-good or fresh demo data, always labelled.
      weather = this.lastGoodWeather || this.demo.getWeather();
      const demo = this.demo;
      alerts = demo.getAlerts(weather);
      riskMap = demo.getRiskMap(weather);
      statusSnapshot = {
        status: 'DEMO',
        sources: {
          weather: { provider: 'DEMO', status: 'DEMO', last_update: weather.timestamp },
          base_map: { provider: 'OpenStreetMap', status: 'LIVE' },
          risk: { provider: 'Platform Risk Engine', status: 'LIVE' },
        },
        server_time: isoNow(),
      };
    }

    const isBackendLive = backendReachable && weather && weather.status === 'LIVE';
    const ageMin = weather && weather.data_age_minutes != null ? weather.data_age_minutes : 0;
    this.data = {
      ...this._initial(),
      mode: isBackendLive ? 'LIVE' : 'DEMO',
      status: isBackendLive ? 'LIVE' : 'DEMO',
      weather,
      alerts: alerts || this.demo.getAlerts(weather),
      riskMap,
      statusSnapshot,
      lastUpdate: weather ? weather.timestamp : isoNow(),
      nextUpdate: new Date(now + CONFIG.refreshIntervalSeconds * 1000).toISOString(),
      stale: ageMin > CONFIG.staleAfterMinutes,
      backendReachable,
      error: backendReachable ? null : 'Backend unreachable — showing simulated data.',
    };
    this._emit();
  }

  start() {
    if (this.running) return;
    this.running = true;
    this._refresh();
    this.timer = setInterval(() => this._refresh(), CONFIG.refreshIntervalSeconds * 1000);
  }

  stop() {
    if (this.timer) clearInterval(this.timer);
    this.timer = null;
    this.running = false;
  }

  getConfig() {
    return { ...CONFIG };
  }

  formatClock(iso) {
    return fmtClock(iso);
  }
}

export const liveService = new LiveService();
export { fmtClock };

File: backend/app/services/live_service.py
"""
Live / near-real-time data service.

Aggregates the current weather observation from the configured provider, derives
active alerts, recomputes settlement risk using live rainfall as an input to the
existing risk engine, and exposes GeoJSON for the live risk map.

Every payload carries source / timestamp / status / data_age_minutes so the UI can
truthfully label LIVE vs DEMO data and never overclaim.
"""
from __future__ import annotations

import time
from collections import deque
from datetime import datetime, timezone, timedelta
from typing import Dict, Any, Optional

from fastapi import Depends
from sqlalchemy.orm import Session

from app.config import settings
from app.database import get_db
from app.models.settlement import Settlement
from app.services.weather_providers import DemoProvider, get_weather_provider
from app.services.risk_engine import evaluate_settlement_risk

IST = timezone(timedelta(hours=5, minutes=30))

# In-memory cache of the last successful observation (failure handling).
_last_good: Dict[str, Any] = {}

# Bounded per-settlement history of platform risk recalculation. The trend arrow is
# derived from ACTUAL successive polls (observation -> risk recalculation), never
# fabricated as a prediction.
_RISK_HISTORY: Dict[str, deque] = {}


def _record_risk(settlement_id: Any, ts: str, rainfall: Optional[float], score: float) -> str:
    """Append this recalculation; return the trend vs the previous actual poll."""
    hist = _RISK_HISTORY.setdefault(str(settlement_id), deque(maxlen=8))
    hist.append({"ts": ts, "rainfall_24h_mm": rainfall, "risk_score": score})
    if len(hist) < 2:
        return "stable"
    prev = hist[-2]["risk_score"]
    if score - prev > 0.5:
        return "increasing"
    if score - prev < -0.5:
        return "decreasing"
    return "stable"


def _now_ist() -> datetime:
    return datetime.now(IST)


def _iso_ist(dt: Optional[datetime] = None) -> str:
    dt = dt or _now_ist()
    return dt.isoformat(timespec="seconds")


def get_live_weather() -> Dict[str, Any]:
    """Current weather observation with TTL caching + last-good failure fallback.

    Within LIVE_REFRESH_INTERVAL_SECONDS the last successful observation is reused
    (coalescing the parallel /api/live/* fetches into one external call per window).
    On provider failure the last good observation is returned as STALE; without any
    cache a clearly-labelled simulated fallback keeps the map alive.
    """
    now = time.time()
    cached = _last_good.get("weather")

    ttl = settings.LIVE_REFRESH_INTERVAL_SECONDS
    if cached and (now - cached["at"]) < ttl:
        payload = dict(cached["payload"])
        payload["data_age_minutes"] = round((now - cached["at"]) / 60.0, 1)
        return payload

    provider = get_weather_provider()
    try:
        obs = provider.get_weather()
    except Exception:
        obs = None

    if obs and obs.get("status") in ("LIVE", "DEMO"):
        _last_good["weather"] = {"payload": obs, "at": time.time()}
        obs["timestamp"] = _iso_ist()
        obs["data_age_minutes"] = 0
        return obs

    # Feed unavailable: fall back to last successful observation (or demo).
    if settings.LIVE_FALLBACK_TO_DEMO and _last_good.get("weather"):
        cached = _last_good["weather"]
        payload = dict(cached["payload"])
        payload["status"] = "STALE"
        payload["data_age_minutes"] = round((time.time() - cached["at"]) / 60.0, 1)
        payload["note"] = "Using last successful observation because the feed is unavailable."
        return payload

    if settings.LIVE_FALLBACK_TO_DEMO:
        # No last-good cached: fall back to clearly-labelled simulated values so the
        # rainfall layer keeps rendering numbers while the feed is down.
        obs = DemoProvider().get_weather()
        obs["status"] = "STALE"
        obs["note"] = "Feed unavailable — showing simulated fallback data."
        return obs

    return {
        "source": "unknown",
        "status": "UNAVAILABLE",
        "timestamp": _iso_ist(),
        "data_age_minutes": 0,
        "note": "Weather feed unavailable. Risk based on latest available data.",
    }


def get_live_alerts(weather: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    """Derives active alerts from the current weather observation.

    No government warning is invented: an alert is only raised from threshold logic
    applied to the current (live or demo) observation, and is clearly labelled with
    its source and severity.
    """
    weather = weather or get_live_weather()
    rain = weather.get("rainfall_24h_mm")

    alerts = []
    severity = "LOW"
    if rain is not None and rain >= 140:
        severity = "HIGH"
        alerts.append(
            {
                "id": "imd-heavy-rain",
                "type": "HEAVY RAIN WARNING",
                "severity": "HIGH",
                "area": "Wayanad",
                "issued": weather.get("timestamp"),
                "message": "24h rainfall crossing 140mm threshold — high-risk slope zones.",
                "source": weather.get("source"),
            }
        )
    elif rain is not None and rain >= 90:
        severity = "MODERATE"
        alerts.append(
            {
                "id": "imd-rain-advisory",
                "type": "RAIN ADVISORY",
                "severity": "MODERATE",
                "area": "Wayanad",
                "issued": weather.get("timestamp"),
                "message": "Elevated rainfall observed — monitor landslide-prone settlements.",
                "source": weather.get("source"),
            }
        )

    status = "LIVE" if weather.get("status") in ("LIVE", "DEMO") else "STALE"
    return {
        "alerts": alerts,
        "count": len(alerts),
        "severity": severity,
        "status": status,
        "source": weather.get("source"),
        "timestamp": _iso_ist(),
        "data_age_minutes": weather.get("data_age_minutes", 0),
    }


def _rain_delta_impact(rainfall_24h_mm: float) -> float:
    """Returns a 0..1 factor reflecting current rainfall pressure on hazard."""
    if rainfall_24h_mm is None:
        return 0.0
    if rainfall_24h_mm >= 140:
        return 0.35
    if rainfall_24h_mm >= 90:
        return 0.2
    if rainfall_24h_mm >= 50:
        return 0.1
    return 0.0


def get_live_risk_map(db: Session = Depends(get_db)) -> Dict[str, Any]:
    """Recomputes settlement risk using live rainfall, returns GeoJSON FeatureCollection."""
    weather = get_live_weather()
    rain = weather.get("rainfall_24h_mm")
    rain_factor = _rain_delta_impact(rain) if rain is not None else 0.0

    settlements = db.query(Settlement).all()
    features = []
    now_ts = _iso_ist()
    for s in settlements:
        st_dict = {
            "landslide_risk": s.landslide_risk,
            "rainfall_risk": min(100.0, s.rainfall_risk + rain_factor * 100.0),
            "slope_risk": s.slope_risk,
            "historical_risk": s.historical_risk,
            "housing_vulnerability": s.housing_vulnerability,
            "households": s.households,
            "priority_households": s.priority_households,
        }
        score, level, components, factors, summary = evaluate_settlement_risk(st_dict)
        hist = _RISK_HISTORY.get(str(s.id))
        prev_poll = hist[-1]["risk_score"] if hist else None
        trend = _record_risk(s.id, now_ts, rain, score)
        features.append(
            {
                "type": "Feature",
                "properties": {
                    "id": s.id,
                    "name": s.name,
                    "risk_score": round(score, 1),
                    "risk_level": level,
                    "households": s.households,
                    "priority_households": s.priority_households,
                    "previous_risk_score": round(prev_poll, 1) if prev_poll is not None else round(s.risk_score, 1),
                    "previous_risk_level": s.risk_level,
                    "rainfall_24h_mm": rain,
                    "observed_at": weather.get("timestamp"),
                    "trend": trend,
                },
                "geometry": {"type": "Point", "coordinates": [s.longitude, s.latitude]},
            }
        )

    return {
        "updated_at": _iso_ist(),
        "risk_calculated_at": _iso_ist(),
        "source": weather.get("source"),
        "status": weather.get("status"),
        "data_age_minutes": weather.get("data_age_minutes", 0),
        "rainfall_24h_mm": rain,
        "rain_impact_factor": round(rain_factor, 2),
        "type": "FeatureCollection",
        "features": features,
    }


def get_live_status() -> Dict[str, Any]:
    """Aggregated health snapshot of all live feeds."""
    weather = get_live_weather()
    last_at = _last_good.get("weather", {}).get("at")
    age = 0
    if last_at:
        age = round((time.time() - last_at) / 60.0, 1)

    stale = age > settings.LIVE_STALE_AFTER_MINUTES
    weather_status = weather.get("status")  # LIVE | DEMO | STALE | UNAVAILABLE
    root_status = "STALE" if stale or weather_status == "STALE" else weather_status
    return {
        "status": root_status,
        "sources": {
            "weather": {
                "provider": weather.get("source"),
                "status": weather.get("status"),
                "last_update": weather.get("timestamp"),
                "data_age_minutes": weather.get("data_age_minutes", age),
            },
            "base_map": {"provider": "OpenStreetMap", "status": "LIVE"},
            "risk": {"provider": "Platform Risk Engine", "status": "CALCULATED", "last_update": weather.get("timestamp")},
        },
        "refresh_interval_seconds": settings.LIVE_REFRESH_INTERVAL_SECONDS,
        "stale_after_minutes": settings.LIVE_STALE_AFTER_MINUTES,
        "server_time": _iso_ist(),
    }